File: haskell/Bindings.hs
module Bindings where

import Text.XML.Light
import Data.Map (Map, insert, lookup, member)
import Data.Maybe (fromMaybe)
import System.FilePath
import Control.Monad (mplus, foldM)
import Text.Regex.Posix
import System.Posix.Files (fileExist, createSymbolicLink, setFileMode)
import Text.JSON (encode)

import Support
import Config
import Selections
import Command (buildCommand)
import Basedir (save, cache)

data WhichEnd = Prepend | Append deriving Show

data AddMode = AddMode { pos :: WhichEnd
		       , defaultValue :: Maybe String
		       , separator :: String
		       } deriving Show

data Mode = Add AddMode | Replace deriving Show

data EnvSource = InsertPath FilePath | Value String deriving Show

data ExecBindingType = InPath | InVar deriving Show

data Binding = EnvironmentBinding VarName Mode EnvSource
	     | ExecutableBinding ExecBindingType String String
	     deriving Show

processEnvironmentBinding :: Element -> Binding
processEnvironmentBinding bindingElem = EnvironmentBinding name mode bindingSource
	where attr attrName = findAttr (mkQName attrName) bindingElem
	      Just name = attr "name"
	      mode = case fromMaybe "prepend" (attr "mode") of
	      	"prepend" -> Add $ AddMode Prepend (attr "default") (fromMaybe [searchPathSeparator] $ attr "separator")
	      	"append" -> Add $ AddMode Append (attr "default") (fromMaybe [searchPathSeparator] $ attr "separator")
		"replace" -> Replace
		m -> error $ "Bad mode " ++ m
	      bindingSource = case (attr "insert", attr "value") of
				  (Nothing, Nothing) -> error "Missing 'insert' or 'value'"
				  (Just i, Nothing) -> InsertPath i
				  (Nothing, Just v) -> Value v
				  (Just _, Just _) -> error "Can't use 'insert' and 'value' together"

processExecutableBinding :: ExecBindingType -> Element -> Binding
processExecutableBinding bindingType bindingElem = ExecutableBinding bindingType name commandName
	where name = requireAttr "name" bindingElem
	      commandName = fromMaybe "run" (findAttr (mkQName "command") bindingElem)

processBinding :: String -> Element -> [Binding]
processBinding "environment" e = [processEnvironmentBinding e]
processBinding "executable-in-path" e = [processExecutableBinding InPath e]
processBinding "executable-in-var" e = [processExecutableBinding InVar e]
processBinding _ _ = []

-- Find all the bindings, in document order
-- Excludes bindings to unselected optional components
collectBindings :: Selections -> [(InterfaceURI, Binding)]
collectBindings sels = do sel <- filterChildrenName (hasZName "selection") (root sels)
                          let iface = requireAttr "interface" sel
                          getBindings True True iface sel
    where getBindings deps commands iface parent =
            do (name, child) <- ziChildren parent
               case name of
                    "command" | commands -> getBindings True False iface child
                    x | (x == "requires" || x == "runner") && deps ->
                             let depIface = requireAttr "interface" child in
                             if depIface `member` (selections sels)
                               then getBindings False False depIface child
                               else []  -- Optional dependency which was not selected
                    _ -> do binding <- processBinding name child
                            [(iface, binding)]

join :: WhichEnd -> String -> Maybe String -> String -> String
join _ _ Nothing new = new
join Prepend sep (Just old) new = new ++ sep ++ old
join Append sep (Just old) new = old ++ sep ++ new

standardDefault :: VarName -> Maybe String
standardDefault "PATH" = Just "/bin:/usr/bin"
standardDefault "XDG_CONFIG_DIRS" = Just "/etc/xdg"
standardDefault "XDG_DATA_DIRS" = Just "/usr/local/share:/usr/share"
standardDefault _ = Nothing

doEnvBinding :: Maybe FilePath -> Env -> Binding -> Env
doEnvBinding mPath env (EnvironmentBinding name mode bindingSource) =
    case (mPath, bindingSource) of
	    (_, Value v) -> use v
	    (Nothing, InsertPath _) -> env
	    (Just implPath, InsertPath i) -> use $ implPath </> i
    where use newValue = insert name (add newValue) env
          add newValue = case mode of
	    Replace -> newValue
	    Add AddMode {pos = whichEnd, defaultValue = def, separator = sep} ->
		join whichEnd sep oldValue newValue
		where oldValue = (Data.Map.lookup name env) `mplus`
				 def `mplus` (standardDefault name)
doEnvBinding _ env _ = env

validateName :: String -> String
validateName name = if name =~ "^[^./'][^/']*$" then name
		    else error $ "Invalid command name: " ++ name

checkRunenv :: Config -> IO ()
checkRunenv config = do execDir <- save ("0install.net" </> "injector") (cache $ basedirs config)
			let execPath = execDir </> "runenv.haskell"
			x <- fileExist execPath
			if x then return ()
			else createSymbolicLink ((resourceDir config) </> "Runenv") execPath

ensureLauncher :: Config -> String -> IO FilePath
ensureLauncher config name = do execDir <- save ("0install.net" </> "injector" </> "executables" </> name) (cache $ basedirs config)
				let execPath = execDir </> name
				x <- fileExist execPath
				if x then return execDir
				else do createSymbolicLink "../../runenv.haskell" execPath
				        setFileMode execDir 0o500
					return execDir

doExecBindings :: Config -> Selections -> Map InterfaceURI FilePath -> Env -> [(InterfaceURI, Binding)] -> IO Env
doExecBindings config sels pathMap initEnv bindings = checkRunenv config >> foldM doExecBinding initEnv bindings
    where doExecBinding :: Env -> (InterfaceURI, Binding) -> IO Env
	  doExecBinding env (iface, ExecutableBinding bindingType name uncheckedCommandName) =
	    do execDir <- ensureLauncher config name
	       let commandJSON = encode (buildCommand sels env pathMap iface commandName)
	       let env2 = insert ("0install-runenv-" ++ name) commandJSON env
	       let env3 = case bindingType of
			     InVar -> insert name (execDir </> name) env2
			     InPath -> insert "PATH" newValue env2
				where oldValue = (Data.Map.lookup "PATH" env2) `mplus`
						 (standardDefault "PATH")
				      newValue = join Prepend [searchPathSeparator] oldValue execDir
	       return env3
	    where commandName = validateName uncheckedCommandName
	  doExecBinding env _ = return env

File: python/bindings.py
import os, json
from os.path import join

import basedirs, command
from common import *

binding_elements = {'environment', 'executable-in-var', 'executable-in-path'}
dep_elements = {'restricts', 'requires', 'runner'}

def validate_exec_name(name):
	if '/' in name or name.startswith('.') or "'" in name:
		raise SafeException("Invalid <executable> name '%s'" % name)

class EnvironmentBinding:
	"""Indicate the chosen implementation using an environment variable."""
	PREPEND = 'prepend'
	APPEND = 'append'
	REPLACE = 'replace'

	defaults = {
		'PATH': '/bin:/usr/bin',
		'XDG_CONFIG_DIRS': '/etc/xdg',
		'XDG_DATA_DIRS': '/usr/local/share:/usr/share',
	}

	def __init__(self, name, insert, default = None, mode = PREPEND, value=None, separator=None):
		"""
		mode argument added in version 0.28
		value argument added in version 0.52
		"""
		self.name = name
		self.insert = insert
		self.default = default
		self.mode = mode
		self.value = value
		if separator is None:
			self.separator = os.pathsep
		else:
			self.separator = separator


	def __str__(self):
		return ("<environ %(name)s %(mode)s %(insert)s %(value)s>") % \
			{'name': self.name, 'mode': self.mode, 'insert': self.insert, 'value': self.value}

	__repr__ = __str__

	def get_value(self, path, old_value):
		"""Calculate the new value of the environment variable after applying this binding.
		@param path: the path to the selected implementation
		@param old_value: the current value of the environment variable
		@return: the new value for the environment variable"""

		if self.insert is not None:
			if path is None:
				return None
			extra = os.path.join(path, self.insert)
		else:
			assert self.value is not None
			extra = self.value

		if self.mode == EnvironmentBinding.REPLACE:
			return extra

		if old_value is None:
			old_value = self.default
			if old_value is None:
				old_value = EnvironmentBinding.defaults.get(self.name, None)
		if old_value is None:
			return extra
		if self.mode == EnvironmentBinding.PREPEND:
			return extra + self.separator + old_value
		else:
			return old_value + self.separator + extra

class ExecutableBinding:
	IN_PATH = "path"
	IN_VAR = "var"

	def __init__(self, exec_type, name, command):
		validate_exec_name(name)
		assert command

		self.exec_type = exec_type
		self.name = name
		self.command = command
	
	def do_exec_binding(self, config, env, impls, iface):
		exec_dir = basedirs.save_path(config.cache, join("0install.net", "injector", "executables", self.name))
		exec_path = join(exec_dir, self.name)
		if not os.path.exists(exec_path):
			os.symlink('../../runenv.py', exec_path)
			os.chmod(exec_dir, 0o500)

		command_argv = command.build_command(impls, iface, self.command, env)
		if self.exec_type == ExecutableBinding.IN_PATH:
			old = env.get('PATH', EnvironmentBinding.defaults['PATH'])
			env['PATH'] = exec_dir + os.pathsep + old
		else:
			env[self.name] = exec_path

		env['0install-runenv-' + self.name] = json.dumps(command_argv)

modes = {
	'prepend': EnvironmentBinding.PREPEND,
	'append': EnvironmentBinding.APPEND,
	'replace': EnvironmentBinding.REPLACE,
}

def collect_bindings(impls, sels):
	bindings = []
	def process(elem, iface, valid_children):
		for child in elem.childNodes:
			if child.uri != ZI.ns: continue
			if child.name not in valid_children: continue

			if child.name == 'command':
				process(child, iface, binding_elements | dep_elements)
			elif child.name in dep_elements:
				dep_iface = child.attrs["interface"]
				if dep_iface in impls:
					process(child, dep_iface, binding_elements)
			elif child.name == 'environment':
				a = child.attrs
				bindings.append((iface, EnvironmentBinding(
					name = a['name'],
					insert = a.get('insert', None),
					default = a.get('default', None),
					mode = modes[a.get('mode', 'prepend')],
					value = a.get('value', None),
					separator = a.get('separator', None)
				)))
			elif child.name in ('executable-in-var', 'executable-in-path'):
				a = child.attrs
				bindings.append((iface, ExecutableBinding(
					ExecutableBinding.IN_VAR if child.name == 'executable-in-var' else ExecutableBinding.IN_PATH,
					a['name'],
					command = a.get('command', 'run')
				)))
			else:
				raise Exception(child.name)
	for sel in ZI.children(sels, "selection"):
		process(sel, sel.attrs["interface"], {'command'} | binding_elements | dep_elements)
	return bindings

